File: iosApp/iosApp/Source Files/Application Flows/Home/View Models/ShopViewModel.swift
//
//  ShopViewModel.swift
//  R&D KMM Retail
//

import UIKit
import Foundation
import common

class ShopViewModel: ObservableObject {

    @Published var searchText: String = ""

    @Published var recommendedProducts = [RecommendedProduct]()

    @Published var recentlyViewedProducts = [RecentlyViewedProduct]()

    private let shopQuery: ShopQuery
    private let shopService: ShopService
    private let pool = CommonJobsPool()

    init(
        shopQuery: ShopQuery = Provider.shared.shopQuery,
        shopService: ShopService = Provider.shared.shopService
    ) {
        self.shopQuery = shopQuery
        self.shopService = shopService
        pool.run(collect: shopQuery.collectRecommendedProducts) { [weak self] products in
            guard let self = self else { return }
            self.recommendedProducts = products
        }
        pool.run(collect: shopQuery.collectRecentlyViewedProducts) { [weak self] products in
            guard let self = self else { return }
            self.recentlyViewedProducts = products
        }
        pool.run(collect: shopQuery.collectSearchedProductName) { [weak self] productName in
            guard let self = self else { return }
            self.searchText = productName
        }
        shopService.loadData()
    }

    deinit {
        pool.cancelAll()
    }

    func handleSearchProduct(productName: String) {
        shopService.searchProduct(productName: productName)
        HapticFeedback.vibrate(with: .selection)
    }

    func handleRecommendedProductClick(product: RecommendedProduct) {
        shopService.openProductDetails(productId: product.id)
        HapticFeedback.vibrate(with: .selection)
    }

    func handleRecentlyViewedProductClick(product: RecentlyViewedProduct) {
        shopService.openProductDetails(productId: product.id)
        HapticFeedback.vibrate(with: .selection)
    }
}

File: androidApp/src/main/java/com/netguru/retail/android/shop/ShopViewModel.kt
package com.netguru.retail.android.shop

import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.netguru.commondomain.shop.ShopQuery
import com.netguru.commondomain.shop.ShopService
import com.netguru.commondomain.shop.model.RecentlyViewedProduct
import com.netguru.commondomain.shop.model.RecommendedProduct
import com.netguru.retail.Provider
import com.netguru.retail.android.viewModel.query

class ShopViewModel(
    private val shopQuery: ShopQuery = Provider.shopQuery,
    private val shopService: ShopService = Provider.shopService
) : ViewModel() {

    var recommendedProducts by query(shopQuery::collectRecommendedProducts, emptyList())
    var recentlyViewedProducts by query(shopQuery::collectRecentlyViewedProducts, emptyList())
    var searchText by query(shopQuery::collectSearchedProductName, "")

    init {
        shopService.loadData()
    }

    fun handleSearchProduct(productName: String) {
        shopService.searchProduct(productName)
    }

    fun handleProductClick(product: RecommendedProduct) {
        shopService.openProductDetails(product.id)
    }

    fun handleProductClick(product: RecentlyViewedProduct) {
        shopService.openProductDetails(product.id)
    }
}
